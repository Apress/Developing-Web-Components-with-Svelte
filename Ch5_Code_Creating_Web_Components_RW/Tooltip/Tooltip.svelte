<svelte:options tag="cobalt-tooltip" />

<script>
  import { fade } from "svelte/transition";
  export let id = "tooltip";
  export let label;
  export let tip;
  export let timeout = "400";
  export let showHTML = "false";

  let displayHTML = JSON.parse(showHTML);

  let active = false;
  let enterTrigger;
  let leaveTrigger;

  function handleKeydown(e) {
    if (e.key === "Escape") {
      active = false;
      e.target.blur();
    }
  }

  function handleMouseEnter() {
    enterTrigger = setTimeout(() => {
      active = true;
    }, parseInt(timeout, 0));
  }

  function handleMouseLeave() {
    if (enterTrigger) {
      clearTimeout(enterTrigger);
      enterTrigger = null;
    }
    leaveTrigger = setTimeout(() => {
      active = false;
    }, parseInt(timeout, 0));
  }

  function handleInteraction() {
    if (leaveTrigger) {
      clearTimeout(leaveTrigger);
      leaveTrigger = null;
    }
  }
</script>

<div class="tooltip">
  <button
    aria-describedby={id}
    type="button"
    class="trigger"
    on:click={() => (active = true)}
    on:keydown={handleKeydown}
    on:mouseenter={handleMouseEnter}
    on:mouseleave={handleMouseLeave}
  >
    ?
  </button>
  <div aria-hidden={!active} {id} role="tooltip" aria-label={label}>
    {#if active}
      <div
        transition:fade
        class="content"
        on:mouseenter={handleInteraction}
        on:mouseleave={handleMouseLeave}
      >
        {#if displayHTML}
          {@html tip}
        {:else}
          {tip}
        {/if}
      </div>
    {/if}
  </div>
</div>

<style>
  .tooltip {
    position: relative;
    z-index: 2;
  }

  .trigger {
    padding: 0;
    margin: 0;
    width: 19px;
    height: 19px;
    line-height: 15px;
    font-size: 17px;
    text-align: center;
    background-color: transparent;
    border-radius: 50%;
    border: 3px solid #666666;
    color: #999999;
    cursor: pointer;
    font-weight: bold;
  }

  .content {
    all: initial;
    position: absolute;
    left: 0;
    top: 100%;
    width: 300px;
    margin-top: 10px;
    padding: 10px;
    border-radius: 8px;
    box-shadow: rgba(0, 0, 0, 0.24) 0px 3px 8px;
    font-size: 14px;
    font-family: Arial, Helvetica, sans-serif;
  }

  .trigger:focus {
    outline: 2px solid #000;
  }

  [role="tooltip"]:empty {
    display: none;
  }
</style>
