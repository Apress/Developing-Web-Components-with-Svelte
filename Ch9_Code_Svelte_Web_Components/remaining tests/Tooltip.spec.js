import { render, screen } from '@testing-library/svelte';
import Tooltip from "../src/lib/Tooltip/Tooltip.svelte";

describe("Tests for Tooltip", () => {
  const props = {
		tip: '<p>This is an informational tooltip - to learn more <a href="/tutorial">click here</a></p>',
		showHTML: "false",
		timeout: "400",
		label: "more info",
  };

  it("should render properly", () => {
    const result = render(Tooltip, props );
    expect(() => result).not.toThrow();
  });

  it("should appear when hovering over target", () => {
    const result = render(Tooltip, props );
  });

  it("should not show any markup when showHTML is true", () => {
    const result = render(Tooltip, { ...props, showHTML: "true" });
  });  

  it("get a snapshot of component", () => {
    const tree = render(Tooltip, props);
    expect(tree).toMatchSnapshot();
  })
});