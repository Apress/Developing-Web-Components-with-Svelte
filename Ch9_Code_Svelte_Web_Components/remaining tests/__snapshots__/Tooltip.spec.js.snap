// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for Tooltip get a snapshot of component 1`] = `
{
  "component": Tooltip {
    "$$": {
      "after_update": [],
      "before_update": [],
      "bound": {},
      "callbacks": {},
      "context": Map {},
      "ctx": [
        "tooltip",
        "more info",
        "<p>This is an informational tooltip - to learn more <a href="/tutorial">click here</a></p>",
        false,
        false,
        [Function],
        [Function],
        [Function],
        [Function],
        "400",
        "false",
        [Function],
      ],
      "dirty": [
        -1,
      ],
      "fragment": {
        "c": [Function],
        "d": [Function],
        "i": [Function],
        "l": [Function],
        "m": [Function],
        "o": [Function],
        "p": [Function],
      },
      "not_equal": [Function],
      "on_destroy": [
        [Function],
      ],
      "on_disconnect": [],
      "on_mount": [],
      "props": {
        "id": 0,
        "label": 1,
        "showHTML": 10,
        "timeout": 9,
        "tip": 2,
      },
      "root": <div>
        <div
          class="tooltip svelte-1v36vsu"
        >
          <button
            aria-describedby="tooltip"
            class="trigger svelte-1v36vsu"
            type="button"
          >
            ?
          </button>
           
          <div
            aria-hidden="true"
            aria-label="more info"
            class="svelte-1v36vsu"
            id="tooltip"
            role="tooltip"
          />
        </div>
      </div>,
      "skip_bound": false,
      "update": [Function],
    },
    "$$set": [Function],
    "$capture_state": [Function],
    "$inject_state": [Function],
  },
  "container": <body>
    <div>
      <div
        class="tooltip svelte-1v36vsu"
      >
        <button
          aria-describedby="tooltip"
          class="trigger svelte-1v36vsu"
          type="button"
        >
          ?
        </button>
         
        <div
          aria-hidden="true"
          aria-label="more info"
          class="svelte-1v36vsu"
          id="tooltip"
          role="tooltip"
        />
      </div>
    </div>
  </body>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
